### PR TITLE
Implementation of a `Container` struct

### PR DESCRIPTION
**Pull request by thomasgouveia**
As `run0` is quite functional in the actual state, I think we can improve the code by creating a `Container` implementation. (At the moment, simply a file, but give me your opinion about creating a crate for it pls, @sameo :) )

With that, run0 would became a CLI in the future in order to manage our containers.

### CODE REVIEW

**Review by sameo — COMMENTED** [review]
Thanks a lot @thomasgouveia , I really like it. Please go ahead.

**Comment by thomasgouveia** [comment]
Hello @sameo,

Normally the PR is ready.

I tested it, the provided code produces the same behavior than before the refactor :D

Let me know if you have any suggestions, I have a lot of ideas to improve run0 but I need this PR as code base. 

Thank you !

(PS: I added a `Makefile` to provide helpers to run & bundle, and a `.gitignore`)

File: run0/Makefile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+TMP_BUNDLE ?= /tmp/run0-bundle
+
+# Helper to create a bundle and move it into /tmp.
+# Simply call `make bundle` before running `make run`
+bundle:
+	cd ../do-vmm/rootfs && ./mkbundle.sh
```
Comment (sameo) [inline run0/Makefile]: You could have one additional commit that would improve `mkbundle.sh` and take a path as its optional first argument.

File: run0/Makefile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+	mv ../do-vmm/rootfs/ctr-bundle $(TMP_BUNDLE)
+
+# Helper to run run0.
+# Requires that `make bundle` was executed before.
+run:
```
Comment (sameo) [inline run0/Makefile]: I'd prefer if we split that one in 2 targets:

```makefile

build:
    cargo build

run: build
    sudo ./target/debug/....

```

File: run0/src/container/mod.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            .map_or(bundle.clone().join(OCI_RUNTIME_SPEC_ROOTFS), |x| {
+                bundle.clone().join(x.path().to_path_buf())
+            });
+
+        // Get the container namespaces if the linux block is defined into the specification.
+        let namespaces = if let Some(linux) = spec.linux() {
+            Namespaces::from(linux.namespaces())
+        } else {
+            Namespaces::default()
+        };
```
Comment (sameo) [inline run0/src/container/mod.rs]: You can follow the same `map_or` pattern as the one we use for rootfs.

File: run0/src/container/mod.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+        let environment = Environment::from(spec.process());
+
+        Ok(Container {
+            bundle: bundle.clone(),
+            rootfs: rootfs.clone(),
+            mounts: Mounts::default(),
+            environment,
```
Comment (sameo) [inline run0/src/container/mod.rs]: nit:

```rust
environment: Environment::from(spec.process());
```

File: run0/src/container/mod.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        } else {
+            Namespaces::default()
+        };
+
+        let environment = Environment::from(spec.process());
+
+        Ok(Container {
+            bundle: bundle.clone(),
+            rootfs: rootfs.clone(),
+            mounts: Mounts::default(),
```
Comment (sameo) [inline run0/src/container/mod.rs]: You can skip that one and use `..Default::default()` at the end of the `Container` initialization:

```rust
Ok(Container {
             bundle: bundle.clone(),
             rootfs: rootfs.clone(),
             environment: Environment::from(spec.process()),
             namespaces,
             ..Default::default()
}            
```

File: run0/src/container/mod.rs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /// Run the container.
+    pub fn run(&self) -> Result<()> {
+        let mounts = self.mounts.clone();
+        let code = unsafe {
+            unshare::Command::new("/bin/sh")
+                .chroot_dir(&self.rootfs)
+                .unshare(&*self.namespaces.get())
+                .pre_exec(move || Mounts::apply(&mounts))
```
Comment (sameo) [inline run0/src/container/mod.rs:85]: Oh, that's neat :-)

**Review by sameo — COMMENTED** [review]
PR looks good, thanks for the cleanup! I only have a few minor comments to address.

2 things:

1. Please fix the linter warnings
2. Could you please remove the `(feat)` prefix on your commit? Just do `run0: Isolate...`

**Comment by thomasgouveia** [comment]
@sameo , Is it ok for you now ?

**Review by sameo — APPROVED** [review]

**Comment by thomasgouveia** [comment]
Nightly and beta check was broken, this should be fixed with my last commit

**Pull request merged**